### ISSUE TITLE
ENH: Add index_as_fid option to write_dataframe?

### ISSUE THREAD

**Issue by theroggy** [geopandas/pyogrio#287]
There is a `fid_as_index` option in `read_dataframe` to be able to... read the `fid`. However, there is no explicit option to e.g. ìndex_as_fid` in `write_dataframe` to write it again.

There is a workaround: if you add the index as a column named "fid" to the dataframe you will write the `fid`, at least for Geopackages. 

For e.g. shapefiles writing fid's is not relevant as they don't save fid's.

Not sure if an explicit option is needed... an alternative could be to just document the "workaround".

**Comment by jorisvandenbossche** [comment]
Do you know if GDAL exposes some information or metadata about this? For example so that we can raise an error if the file format actually does not support it?

**Comment by theroggy** [comment]
> Do you know if GDAL exposes some information or metadata about this? For example so that we can raise an error if the file format actually does not support it?

Yes, if [OGR_L_GetFIDColumn()](https://gdal.org/doxygen/classOGRLayer.html#af0cb9a05310d7b17d9dfec9e83a18e61)  returns "", the file type cannot save `fid`'s. If it returns another string, that's the column name used to store the `fid`. 

Examples: 
- for shapefile "" is returned. 
- for Geopackage "fid" is returned: that's the default name of the column the fid is stored in.
- for sqlite files I never tested/used it, but according to the gdal doc the default column name for the `fid` is "OGC_FID", so I suppose that `OGR_L_GetFIDColumn()` will return that and I suppose that you'll have to call the column like that in the GeoDataFrame if you want the `fid`'s to be retained.

**Comment by jorisvandenbossche** [comment]
Nice, in that case this sounds as a good option to me.

**Comment by theroggy** [comment]
I just found another way to save the index of the GeoDataFrame as "fid" (primary key) in a GPKG: if the name of the index is "fid" it will also be saved as "fid" in the file.

xref: https://github.com/geopandas/geopandas/issues/3217

**Comment by jorisvandenbossche** [comment]
(moved this to the next milestone)

**Comment by evetion** [comment]
Additionaly, one can pass `fid="custom_index_name"` to the kwargs of write_dataframe, and make any column the primary key in a GPKG. See https://gdal.org/drivers/vector/gpkg.html#layer-creation-options.

Ideally we would then also set the correct index name on reading when `fid_as_index` is True, instead of [hardcoding](https://github.com/geopandas/pyogrio/blob/4aea0909c7371db9528ed999689e672f7af6f0da/pyogrio/geopandas.py#L316C39-L316C42) it to "fid". `OGR_L_GetFIDColumn` should give us the name to use here.
